test/kad-dht.js

```diff
@@ -3,110 +3,516 @@
 chai.use(dirtyChai)
+chai.use(chaiBytes)
 
 const series = require('async/series')
 const crypto = require('crypto')
 const parallel = require('async/parallel')
 const waterfall = require('async/waterfall')
-const bl = require('bl')
 
 const DaemonFactory = require('ipfsd-ctl')
 const goDf = DaemonFactory.create()
 const jsDf = DaemonFactory.create({ type: 'js' })
 
-describe.skip('kad-dht', () => {
-  describe('a JS node in the land of Go', () => {
-    let jsD
-    let goD1
-    let goD2
-    let goD3
+const getConfig = (bootstrap) => ({
+  Bootstrap: bootstrap,
+  Discovery: {
+    MDNS: {
+      Enabled: false
+    },
+    webRTCStar: {
+      Enabled: false
+    }
+  }
+})
+
+const spawnGoDaemon = (bootstrap = [], callback) => {
+  goDf.spawn({
+    initOptions: { bits: 1024 },
+    config: getConfig(bootstrap)
+  }, callback)
+}
+
+const spawnJsDaemon = (bootstrap = [], callback) => {
+  jsDf.spawn({
+    initOptions: { bits: 512 },
+    config: getConfig(bootstrap),
+    args: ['--enable-dht-experiment']
+  }, callback)
+}
+
+const getNodeId = (node, callback) => {
+  node.api.id((err, res) => {
+    expect(err).to.not.exist()
+    expect(res.id).to.exist()
+
+    callback(null, res.addresses[0])
+  })
+}
+
+describe.only('kad-dht', () => {
+  describe.only('a JS network', () => {
+    let bootstrapAddr
+    let node0
+    let node1
+    let node2
+    let node3
+    let data
+
+    // spawn bootstrap daemon and get address
+    before(function (done) {
+      this.timeout(60 * 1000)
+
+      spawnJsDaemon([], (err, node) => {
+        expect(err).to.not.exist()
+        node0 = node
+
+        getNodeId(node0, (err, res) => {
+          expect(err).to.not.exist()
+          bootstrapAddr = res
+          done()
+        })
+      })
+    })
+
+    // spawn daemons
+    before(function (done) {
+      this.timeout(70 * 1000)
+      const bootstrap = [bootstrapAddr]
 
-    before((done) => {
       parallel([
-        (cb) => goDf.spawn({ initOptions: { bits: 1024 } }, cb),
-        (cb) => goDf.spawn({ initOptions: { bits: 1024 } }, cb),
-        (cb) => goDf.spawn({ initOptions: { bits: 1024 } }, cb),
-        (cb) => jsDf.spawn({ initOptions: { bits: 512 } }, cb)
+        (cb) => spawnJsDaemon(bootstrap, cb),
+        (cb) => spawnJsDaemon(bootstrap, cb),
+        (cb) => spawnJsDaemon(bootstrap, cb)
       ], (err, nodes) => {
         expect(err).to.not.exist()
-        goD1 = nodes[0]
-        goD2 = nodes[1]
-        goD3 = nodes[2]
-        jsD = nodes[3]
+        node1 = nodes[0]
+        node2 = nodes[1]
+        node3 = nodes[2]
         done()
       })
     })
 
-    after((done) => {
+    // create data
+    before(function (done) {
+      data = crypto.randomBytes(9001)
+      done()
+    })
+
+    after(function (done) {
+      this.timeout(100 * 1000)
       series([
-        (cb) => goD1.stop(cb),
-        (cb) => goD2.stop(cb),
-        (cb) => goD3.stop(cb),
-        (cb) => jsD.stop(cb)
+        (cb) => node1.stop(cb),
```

Comment:
I think this is fine, but it could be even cleaner if all the nodes were in an array.